### PR TITLE
Unify toLower implementations

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

I found 3 instances where different files needed to transform a string into all lower case letters. Two implementations were identical and a 3rd used a standard algorithm. 3 reimplementations of the same thing seemed like justification to create a new private abstraction to handle this. Following patterns I saw elsewhere, I added a `priv` namespace to String.hpp within which `sf::priv::toLower` is implemented. 

## Tasks

* [ ] Tested on Linux
* [ ] Tested on Windows
* [x] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

### CODE REVIEW

File: include/SFML/System/String.hpp
Comment (kimci86) [inline include/SFML/System/String.hpp:46]:
```suggestion
            c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
```
See https://en.cppreference.com/w/cpp/string/byte/tolower#Notes

**Review by kimci86 — COMMENTED** [review]
`sf::priv::toLower` can be simplified into `priv::toLower` or `toLower` when used in the corresponding namespace.

File: include/SFML/System/String.hpp
Comment (ChrisThrasher) [reply include/SFML/System/String.hpp]: Thanks. Fixed.

File: src/SFML/Network/Http.cpp
Comment (vittorioromeo) [inline src/SFML/Network/Http.cpp:30]: `String.hpp` is one of the most expensive SFML headers to compile. We are adding it here with the only purpose of exposing `toLower` -- I am not sure that is a good idea in the context of compilation times.

Comment (Bromeon) [reply src/SFML/Network/Http.cpp:30]: I agree. Chances are there are more such utilities like `sf::toLower()`, and those could maybe combined in an own header `System/Utils.hpp` (if dependency-heavy, with separate `.cpp` file)?

Comment (ChrisThrasher) [reply src/SFML/Network/Http.cpp:30]: @vittorioromeo I totally agree. I originally wanted to put this in a private header somewhere in src/ since this is only needed in .cpp files. However, I saw no internal generic utilities file/library and I saw that many files included a `priv::` namespace in public headers for things like this so I copied that pattern.

@Bromeon I think a Utils.[hpp|cpp] is a great idea. There ought to be a central location for common implementation details like this so we can encourage better code reuse.

Comment (Bromeon) [reply src/SFML/Network/Http.cpp:30]: The only problem I see is that such "utils" tend to aggregate all kinds of unrelated functionality, which often comes with various different `#include`s, ultimately increasing compilation time for each of the client files.

For example, I was thinking this might be a place for the min/max functionality used in `Rect` and `Vector2/3`, but that would mean those need to include `<string>`, which is a no-no.

We can still start with `Utils`, and rename it as we see fit.

Comment (vittorioromeo) [reply src/SFML/Network/Http.cpp:30]: I honestly would also be fine with something like `Utils/ToLower.hpp`, and have `Utils` as a folder.

Comment (Bromeon) [reply src/SFML/Network/Http.cpp:30]: Also a possibility. In case of small dependencies (here `<cctype>` which might already be included by `<string>`), we can also abandon the .cpp file entirely.

Comment (ChrisThrasher) [reply src/SFML/Network/Http.cpp:30]: What if we put private headers in src/? That way we can avoid `priv::` and `Priv/` and have a much stronger guarantee that no user will depend on these utilities.

**Comment by ChrisThrasher** [comment]
This latest revision moves this utility out of String.hpp and into System/Priv/Utils.hpp. This is good because it means we don't need to bring in the entire String.hpp header when these 3 files just need this one utility. I think we can do better though. Because we ideally want these private implementation details to be unobservable to users, I think the ideal solution is to move this file into src/. I'm still considering how exactly to implement that in a way that is more granular than simply adding 

```cmake
include_directories(.)
```

to src/CMakeLists.txt. This works, but it means that the System module could, for example, depend on files from Network without having to express a dependency on that module which we probably don't want to allow. We can't add any new public include paths inside src/ since those will be visible to users so we might be forced to do something like that `include_directories(.)` line to allow sharing files between .cpp files without them being visible to users.

**Comment by ChrisThrasher** [comment]
https://github.com/ChrisThrasher/SFML/commit/c8e963243c2e3feb9f12dd63c300c671428530d6

Here's one way we could potentially add private headers to src/ to prevent having to add more implementation details to public headers.

**Comment by kimci86** [comment]
@ChrisThrasher It is already possible to put implementation headers in `src/` without new CMake commands.
See https://github.com/SFML/SFML/blob/master/cmake/Macros.cmake#L212-L214

**Comment by ChrisThrasher** [comment]
> @ChrisThrasher It is already possible to put implementation headers in `src/` without new CMake commands.
> See https://github.com/SFML/SFML/blob/master/cmake/Macros.cmake#L212-L214

Oh, lovely. In that case I feel ever better about moving this function to src/SFML/System/Utils.hpp (or maybe ToLower.hpp). I merged that commit into this PR. I'll squash commits as requested once we all agree on a solution. I'll keep both commits for now in case we choose to remove the latest commit I added.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Audio/SoundFileWriterWav.cpp
Comment (vittorioromeo) [inline src/SFML/Audio/SoundFileWriterWav.cpp:31]: Can you check if this include can be removed now? `<algorithm>` is quite expensive as a header, would be nice to get rid of it as part of this PR if it's unnecessary.

**Review by vittorioromeo — CHANGES_REQUESTED** [review]
One minor thing, LGTM otherwise

File: src/SFML/Audio/SoundFileWriterWav.cpp
Comment (ChrisThrasher) [reply src/SFML/Audio/SoundFileWriterWav.cpp:31]: Looks like it can. I'll remove it.

File: src/SFML/System/Utils.hpp
Comment (roughbits01) [inline src/SFML/System/Utils.hpp:37]: Is there a reason `str` is passed by copy? (I know it was like this before your refactor)

Comment (ChrisThrasher) [reply src/SFML/System/Utils.hpp]: A copy is required if we want to modify the argument then return it.

Comment (roughbits01) [reply src/SFML/System/Utils.hpp]: Oh that makes sense! 👍

**Comment by ChrisThrasher** [comment]
@vittorioromeo I found two more files with custom `toLower` implementations and was able to remove `<algorithm>` in two more places!

**Comment by ChrisThrasher** [comment]
```
  D:\a\SFML\SFML\src\SFML\Audio\SoundFileWriterFlac.cpp(101): error C2039: 'min': is not a member of 'std' [D:\a\SFML\SFML\build\src\SFML\Audio\sfml-audio.vcxproj]
793
  D:\a\SFML\SFML\src\SFML\Audio\SoundFileWriterFlac.cpp(101): error C3861: 'min': identifier not found [D:\a\SFML\SFML\build\src\SFML\Audio\sfml-audio.vcxproj]
794
  D:\a\SFML\SFML\src\SFML\Audio\SoundFileWriterOgg.cpp(145): error C2039: 'min': is not a member of 'std' [D:\a\SFML\SFML\build\src\SFML\Audio\sfml-audio.vcxproj]
795
  D:\a\SFML\SFML\src\SFML\Audio\SoundFileWriterOgg.cpp(145): error C3861: 'min': identifier not found [D:\a\SFML\SFML\build\src\SFML\Audio\sfml-audio.vcxproj]
796
  D:\a\SFML\SFML\src\SFML\Audio\SoundFileWriterOgg.cpp(150): error C2039: 'min': is not a member of 'std' [D:\a\SFML\SFML\build\src\SFML\Audio\sfml-audio.vcxproj]
797
  D:\a\SFML\SFML\src\SFML\Audio\SoundFileWriterOgg.cpp(150): error C3861: 'min': identifier not found [D:\a\SFML\SFML\build\src\SFML\Audio\sfml-audio.vcxproj]
```

I may have been too quick to remove `<algorithm>`.

@vittorioromeo Don't we have a non-standard min/max implementation that we're using to avoid more inclusions of `<algorithm>`? I think we need to add that to src\SFML\Audio\SoundFileWriterOgg.cpp and src\SFML\Audio\SoundFileWriterFlac.cpp.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
I followed through on @vittorioromeo's recommendation but the PR still shows as being blocked by an unaddressed request. Not sure what to make of that but it appears that everything else is good to go.

**Comment by eXpl0it3r** [comment]
> the PR still shows as being blocked by an unaddressed request

It's shown as such, until the reviewer approves it 😉

**Review by vittorioromeo — APPROVED** [review]
LGTM, thanks!!

**Pull request merged**
